File: detection-tools/floci/floci_exfil.py
"""
Optional Floci S3 mirror for SCAS scenarios (SCAS_FLOCI_ENABLED=1).
Python equivalent of floci-exfil.js — calls scripts/floci-upload-json.sh.
"""

from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path


def find_upload_script() -> Path | None:
    directory = Path.cwd()
    for _ in range(12):
        candidate = directory / "scripts" / "floci-upload-json.sh"
        if candidate.is_file():
            return candidate
        parent = directory.parent
        if parent == directory:
            break
        directory = parent
    return None


def upload_json(
    scenario_id: str,
    key_suffix: str,
    payload: dict,
    prefix: str = "exfil",
) -> bool:
    if os.environ.get("TESTBENCH_MODE") != "enabled":
        return False
    if os.environ.get("SCAS_FLOCI_ENABLED") != "1":
        return False
    script = find_upload_script()
    if script is None:
        return False
    try:
        subprocess.run(
            [str(script), str(scenario_id), key_suffix, prefix],
            input=json.dumps(payload).encode("utf-8"),
            check=True,
            capture_output=True,
            env=os.environ.copy(),
        )
        return True
    except (subprocess.CalledProcessError, OSError):
        return False

File: detection-tools/floci/floci-exfil.js
/**
 * Optional Floci S3 mirror for SCAS scenarios (SCAS_FLOCI_ENABLED=1).
 * Walks up from cwd to find scripts/floci/floci-upload-json.sh (works from victim-app / packages).
 */
'use strict';

const { execFileSync } = require('child_process');
const fs = require('fs');
const path = require('path');

function findUploadScript() {
  let dir = process.cwd();
  for (let i = 0; i < 12; i++) {
    const candidate = path.join(dir, 'scripts', 'floci-upload-json.sh');
    if (fs.existsSync(candidate)) return candidate;
    const parent = path.dirname(dir);
    if (parent === dir) break;
    dir = parent;
  }
  return null;
}

/**
 * @param {string} scenarioId - e.g. '06', '17', '21'
 * @param {string} keySuffix - e.g. 'stage1', 'harvested-credentials'
 * @param {object} payload
 * @param {string} [prefix] - S3 key prefix (default 'exfil', use 'chain' for scenario 17)
 */
function uploadJson(scenarioId, keySuffix, payload, prefix = 'exfil') {
  if (process.env.TESTBENCH_MODE !== 'enabled') return false;
  if (process.env.SCAS_FLOCI_ENABLED !== '1') return false;
  const script = findUploadScript();
  if (!script) return false;
  try {
    execFileSync(script, [String(scenarioId), keySuffix, prefix], {
      input: JSON.stringify(payload),
      stdio: ['pipe', 'pipe', 'pipe'],
      env: process.env,
    });
    return true;
  } catch {
    return false;
  }
}

module.exports = { uploadJson, findUploadScript };
